File: ChanSentry.Cli/Services/ThreadProcessingService.cs
using ChanSentry.Cli.Utils;
using ChanSentry.Common.Models;
using Spectre.Console;

namespace ChanSentry.CLI.Services;

public class ThreadProcessingService
{
    private readonly ThreadFetchService _fetchService;
    private readonly MediaDownloadService _downloadService;

    public ThreadProcessingService()
    {
        _fetchService = new ThreadFetchService();
        _downloadService = new MediaDownloadService();
    }

    public async Task ProcessThreadAsync(WatchedThread thread)
    {
        DisplayThreadCheckMessage(thread);
        
        var fetchResult = await _fetchService.FetchThreadAsync(thread);

        if (fetchResult.IsSuccess && fetchResult.ThreadData != null)
        {
            thread.LastChecked = DateTime.UtcNow;
            await ProcessSuccessfulFetchAsync(thread, fetchResult.ThreadData);
        }
        else if (fetchResult.IsNotModified)
        {
            thread.LastChecked = DateTime.UtcNow;
            DisplayNotModifiedMessage(thread);
        }
        else
        {
            HandleFetchError(thread, fetchResult.StatusCode);
        }
    }

    private async Task ProcessSuccessfulFetchAsync(WatchedThread thread, Common.Models.Thread threadData)
    {
        UpdateThreadSubjectIfNeeded(thread, threadData);
        
        var (mediaPosts, newMedia) = GetMediaPosts(thread, threadData);
        
        await _downloadService.DownloadMediaFilesAsync(newMedia, thread);
        
        thread.TotalDownloadedFiles = mediaPosts.Count;
    }

    private static void UpdateThreadSubjectIfNeeded(WatchedThread thread, Common.Models.Thread threadData)
    {
        if (!string.IsNullOrWhiteSpace(thread.Subject))
            return;

        var subject = threadData.Posts.FirstOrDefault()?.Subject ?? string.Empty;
        thread.Subject = subject;
        AnsiConsole.MarkupLine($"[blue]Retrieved subject: {FileNameSanitizer.EscapeMarkup(subject)}[/]");
    }

    private static (List<Post> mediaPosts, List<Post> newMedia) GetMediaPosts(WatchedThread thread, Common.Models.Thread threadData)
    {
        AnsiConsole.MarkupLine($"[blue]Successfully fetched thread {thread.ThreadId} on /{thread.Board}/[/]");

        var mediaPosts = threadData.Posts.Where(post => post.HasMedia).ToList();
        var newMedia = mediaPosts.Skip(thread.TotalDownloadedFiles).ToList();
        
        AnsiConsole.MarkupLine($"[green]Found {newMedia.Count} new media files in thread {thread.ThreadId}[/]");
        
        return (mediaPosts, newMedia);
    }

    private static void DisplayThreadCheckMessage(WatchedThread thread)
    {
        var displaySubject = string.IsNullOrWhiteSpace(thread.Subject) ? "No Subject" : thread.Subject;
        var escapedSubject = FileNameSanitizer.EscapeMarkup(displaySubject);
        AnsiConsole.MarkupLine($"[grey]Checking: [/]{thread.Board}/{thread.ThreadId} - [green]{escapedSubject}[/]");
    }

    private static void DisplayNotModifiedMessage(WatchedThread thread)
    {
        AnsiConsole.MarkupLine($"[blue]Thread {thread.ThreadId} on /{thread.Board}/ has not been modified since last check.[/]");
    }

    private static void HandleFetchError(WatchedThread thread, System.Net.HttpStatusCode? statusCode)
    {
        thread.ErrorCount++;
        
        AnsiConsole.MarkupLine($"[red]Failed to fetch thread {thread.ThreadId} on /{thread.Board}/ - Status Code: {statusCode} (Total Errors: {thread.ErrorCount}/3)[/]");

        if (thread.ErrorCount >= 3)
        {
            AnsiConsole.MarkupLine($"[red]Thread has failed {thread.ErrorCount} times and will be deleted from Watched Threads list.[/]");
        }
    }
}

File: src/thread-processing-service.js
import { MAX_ERROR_COUNT } from './constants.js';
import { postHasMedia } from './models.js';
import { escapeMarkup } from './file-name-sanitizer.js';
import { ThreadFetchService } from './thread-fetch-service.js';
import { MediaDownloadService } from './media-download-service.js';
import { WatchedThreadService } from './watched-thread-service.js';

export class ThreadProcessingService {
  constructor({
    fetchService = new ThreadFetchService(),
    downloadService = new MediaDownloadService(),
    watchedThreadService = new WatchedThreadService(),
    logger = console
  } = {}) {
    this.fetchService = fetchService;
    this.downloadService = downloadService;
    this.watchedThreadService = watchedThreadService;
    this.logger = logger;
  }

  async processThread(thread, allWatchedThreads) {
    const displaySubject = thread.Subject?.trim() ? thread.Subject : 'No Subject';
    this.logger.log(`Checking: ${thread.Board}/${thread.ThreadId} - ${escapeMarkup(displaySubject)}`);

    const fetchResult = await this.fetchService.fetchThread(thread);

    if (fetchResult.isSuccess && fetchResult.threadData) {
      thread.LastChecked = new Date().toISOString();
      await this.processSuccessfulFetch(thread, fetchResult.threadData);
      return;
    }

    if (fetchResult.isNotModified) {
      thread.LastChecked = new Date().toISOString();
      this.logger.log(`Thread ${thread.ThreadId} on /${thread.Board}/ has not been modified since last check.`);
      return;
    }

    await this.handleFetchError(thread, allWatchedThreads, fetchResult.statusCode);
  }

  async processSuccessfulFetch(thread, threadData) {
    if (!thread.Subject?.trim()) {
      thread.Subject = threadData.Posts[0]?.Subject ?? '';
      this.logger.log(`Retrieved subject: ${escapeMarkup(thread.Subject)}`);
    }

    this.logger.log(`Successfully fetched thread ${thread.ThreadId} on /${thread.Board}/`);

    const mediaPosts = threadData.Posts.filter(postHasMedia);
    const newMedia = mediaPosts.slice(thread.TotalDownloadedFiles);
    this.logger.log(`Found ${newMedia.length} new media files in thread ${thread.ThreadId}`);

    await this.downloadService.downloadMediaFiles(newMedia, thread);
    thread.TotalDownloadedFiles = mediaPosts.length;
  }

  async handleFetchError(thread, allWatchedThreads, statusCode) {
    if (statusCode === 404) {
      this.logger.error(`Thread ${thread.ThreadId} on /${thread.Board}/ returned 404 and will be deleted from Watched Threads list immediately.`);
      const index = allWatchedThreads.indexOf(thread);
      if (index >= 0) {
        allWatchedThreads.splice(index, 1);
      }
      await this.watchedThreadService.saveWatchedThreads(allWatchedThreads);
      return;
    }

    thread.ErrorCount += 1;
    this.logger.error(`Failed to fetch thread ${thread.ThreadId} on /${thread.Board}/ - Status Code: ${statusCode} (Total Errors: ${thread.ErrorCount}/3)`);

    if (thread.ErrorCount >= MAX_ERROR_COUNT) {
      this.logger.error(`Thread has failed ${thread.ErrorCount} times and will be deleted from Watched Threads list.`);
      const index = allWatchedThreads.indexOf(thread);
      if (index >= 0) {
        allWatchedThreads.splice(index, 1);
      }
      await this.watchedThreadService.saveWatchedThreads(allWatchedThreads);
    }
  }
}
